proxy/src/connection.rs

```diff
@@ -122,9 +125,22 @@ impl BoundPort {
                     // libraries don't have the necessary API for that, so just
                     // do it here.
                     set_nodelay_or_warn(&socket);
-                    let connection = Connection::plain(socket);
-                    future::ok((connection, remote_addr))
+                    match tls_config.clone() {
+                        Some(tls_config) => {
+                            Either::A(
+                                tls::Connection::accept(socket, tls_config)
+                                    .map(move |tls| (Connection::new(Box::new(tls)), remote_addr)))
+                        },
+                        None => Either::B(future::ok((Connection::new(Box::new(socket)), remote_addr))),
+                    }
                 })
+                .map(Some)
+                .or_else(|err| {
+                    debug!("error handshaking: {}", err);
+                    let r: Option<(Connection, SocketAddr)> = None;
+                    future::ok(r)
+                })
```

Comment:
```rust
                .map(Some)
                .or_else(|err| {
                    debug!("error handshaking: {}", err);
                    let r: Option<(Connection, SocketAddr)> = None;
                    future::ok(r)
                })
```

I feel like there must be a better way of inserting a "log error and keep failing" step. Probably something like:

```rust
                .or_else(|err| {
                    debug!("error handshaking: {}", err);
                    future::err(err)
                })
```

I'll try that now.

Reply:
Done. It works!